fennel.lua

```diff
@@ -2950,6 +2955,35 @@ that argument name begins with ?."
               (let [(ok view) (pcall require :fennelview)]
                 `(print ,((if ok view tostring)
                           (macroexpand form _SCOPE)))))
+ :import-macros (fn import-macros [...]
+                  "Imports the macros from each macro module in binding/modname pairs.
+Example:
+  (import-macros mymacros                 :proj.macros ; bind all to namespace
+                 {:macro1 alias : macro2} :my-macros)  ; import by name"
+                  (assert (= 0 (% (select :# ...) 2))
+                          "expected even number of binding/modulename pairs")
+                  (for [i 1 (select :# ...) 2]
+                    (local (bindings modname) (select i ...))
+                    ;; generate a subscope of current scope, use require-macros to bring in macro
+                    ;; module. after that, we just copy the macros from subscope to scope.
+                    (local scope (get-scope))
+                    (local subscope (fennel.scope scope))
+                    (fennel.compileString (string.format "(require-macros %q)" modname)
+                                          {:scope subscope})
+                    (if (sym? bindings)
+                        ;; import a whole set of macros to a symbol namespace
+                        (do (tset scope.macros (. bindings 1) {})
+                            (each [k v (pairs subscope.macros)]
+                              (tset (. scope.macros (. bindings 1)) k v)))
+
+                        ;; 1-level table destructuring for importing individual macros
+                        (table? bindings)
+                        (each [macro-name [import-key] (pairs bindings)]
+                          (assert (. subscope.macros macro-name)
```

Comment:
I think you're right that we should assert that type=function here, not just non-nil.

Reply:
I can do that, but what about in `require-macros`? With the way this PR modifies `macroexpand` to check multisyms that resolve to a macro, it would still be possible to export nested macro objects unless we do the sanitizing in `require-macros`.

That doesn't preclude making the change here to check the binding form, just curious if you think I should disallow `{:foo {:bar (fn [] ...) :baz (fn [] ...) }}` in `require-macros` too.

Reply:
Oh, I guess I was just thinking that every leaf node should be a function, which we can't really solve here. I forgot that the single-level limitation applied only to destructuring, not to the overall shape of the module's table. Never mind!

Reply:
Come to think of it, it wouldn't be very hard to make this tail-recursive, which would allow nesting. The question is: do we want to? It does potentially add complexity, but it would allow thinks like

- `macros/foo.fnl`
- `macros/bar.fnl`
- `macros/init.fnl` - returns `{:foo foo-macros :bar bar-macros}`

Reply:
No, I think this would give the impression that there is a relationship between filenames which doesn't actually hold. Different modules that resolve to files in the same directory do not actually interact with each other unless you make them, and we shouldn't introduce a connection that the module system on its own does not.